### PR TITLE
base: Add support for read receipts

### PR DESCRIPTION
**Pull request by zecakeh**
Stores them and keeps them updated. Allows to retrieve them from a room.

### CODE REVIEW

File: matrix_sdk_base/src/store/sled_store/mod.rs [hunk 1]
```diff
@@ -127,12 +129,45 @@ impl EncodeKey for (&str, &str, &str) {
...
+///
+/// The key must have been encoded with the `EncodeKey` trait. `position`
+/// corresponds to the position in the tuple before the key was encoded. If it
+/// wasn't encoded in a tuple, use `0`.
+///
+/// Returns `None` if there is no key at `position`.
+pub fn decode_key_value(key: &[u8], position: usize) -> Option<String> {
+    let values: Vec<&[u8]> = key.split(|v| *v == ENCODE_SEPARATOR).collect();
+
+    if position >= values.len() {
```
Comment (poljar) [inline matrix_sdk_base/src/store/sled_store/mod.rs:164]: I think I prefer a `values.get(position).map(String::from_utf8_lossy)` type of thing here.

**Review by poljar — CHANGES_REQUESTED** [review]
This looks mostly fine but can we please add some tests for this since it involves some non-trivial storing and loading operations.

File: matrix_sdk_base/src/store/memory_store.rs [hunk 2]
```diff
@@ -502,3 +502,111 @@ impl StateStore for MemoryStore {
...
+    use serde_json::json;
+
+    use super::{MemoryStore, StateChanges};
+
+    #[async_test]
+    async fn test_receipts_saving() {
+        let store = MemoryStore::new();
+
+        let room_id = room_id!("!test:localhost");
+        let user_id = user_id!("@rikj:jki.re");
```
Comment (poljar) [inline matrix_sdk_base/src/store/memory_store.rs:523]: Let's just use a fake user id here.

File: matrix_sdk_base/src/store/sled_store/mod.rs [hunk 3]
```diff
@@ -933,4 +930,99 @@ mod test {
             .unwrap()
             .is_some());
     }
+
+    #[async_test]
+    async fn test_receipts_saving() {
+        let store = SledStore::open().unwrap();
+
+        let room_id = room_id!("!test:localhost");
+        let user_id = user_id!("@rikj:jki.re");
```
Comment (poljar) [inline matrix_sdk_base/src/store/sled_store/mod.rs:939]: Same as above.

**Review by poljar — CHANGES_REQUESTED** [review]
Thanks, much better now, let's just replace Erik with a fake Erik with a mustache.

File: matrix_sdk_base/src/store/memory_store.rs (hunk 2 above)
Comment (zecakeh) [reply matrix_sdk_base/src/store/memory_store.rs:523]: Oh I didn't know it was a real one. I took the example from the spec

Comment (poljar) [reply matrix_sdk_base/src/store/memory_store.rs:523]: Hmm, Erik is in the spec, well I guess it doesn't matter that much but let's just use `example`, or `alice` like we usually do.

**Comment by poljar** [comment]
Thanks, merged.

**Pull request merged**